File: backend/app/services/gemini_service.py
import os
import json
import ssl
import urllib.request
import urllib.parse
import re
from app.prompts.travel_prompts import SYSTEM_PROMPT_CHAT, SYSTEM_PROMPT_ITINERARY, SYSTEM_PROMPT_SEARCH_EXTRACTION
from app.utils.logger import get_logger

logger = get_logger("app.services.gemini_service")

def _clean_geographic_entity(val: str | None) -> str | None:
    if not val or not isinstance(val, str):
        return None
    cleaned = val.strip()
    prefix_pattern = r'^(?:destinations?\s+(?:in|near|around|to|close\s+to)|places?\s+(?:in|near|around|to|close\s+to)|trip\s+(?:in|near|around|to)|nearby|near|around|close\s+to|in|to|visit)\s+'
    prev = ''
    while cleaned != prev:
        prev = cleaned
        cleaned = re.sub(prefix_pattern, '', cleaned, flags=re.I).strip()

    cleaned = re.sub(r'\s+(?:places?|destinations?|spots?)$', '', cleaned, flags=re.I).strip()
    if re.match(r'^(?:nearby|near|around|close\s+to|places?|destinations?)$', cleaned, re.I):
        return None
    return cleaned.title() if cleaned else None

class GeminiService:
    def __init__(self):
        self.api_key = (
            os.getenv("GEMINI_API_KEY") or 
            os.getenv("GOOGLE_API_KEY") or 
            os.getenv("VITE_GEMINI_API_KEY")
        )
        if self.api_key:
            logger.info("Gemini AI active (Gemini 2.5 Flash initialized with GEMINI_API_KEY).")
            self.url = f"https://generativelanguage.googleapis.com/v1beta/models/gemini-2.5-flash:generateContent?key={self.api_key}"
        else:
            logger.warning("GEMINI_API_KEY is missing. Using fallback mock travel engine.")
            self.url = None

    async def get_chat_response(self, messages: list) -> str:
        """
        Sends the message history to Gemini 2.5 Flash, or returns a conversational mock reply if API key is missing/fails.
        """
        logger.info(f"Gemini Service: Received chat request with {len(messages)} messages.")
        if self.api_key and self.url:
            try:
                contents = []
                for msg in messages:
                    role = msg.get("role", "user")
                    gemini_role = "model" if role in ("assistant", "model") else "user"
                    contents.append({
                        "role": gemini_role,
                        "parts": [{"text": str(msg.get("content", ""))}]
                    })
                
                payload = {
                    "contents": contents,
                    "systemInstruction": {
                        "parts": [{"text": SYSTEM_PROMPT_CHAT}]
                    },
                    "generationConfig": {
                        "temperature": 0.7,
                        "maxOutputTokens": 1000
                    }
                }

                req_data = json.dumps(payload).encode("utf-8")
                req = urllib.request.Request(
                    self.url,
                    data=req_data,
                    headers={"Content-Type": "application/json"},
                    method="POST"
                )
                ctx = ssl.create_default_context()
                ctx.check_hostname = False
                ctx.verify_mode = ssl.CERT_NONE

                with urllib.request.urlopen(req, context=ctx, timeout=30) as response:
                    resp_data = json.loads(response.read().decode("utf-8"))
                    candidates = resp_data.get("candidates", [])
                    if candidates:
                        text_content = candidates[0].get("content", {}).get("parts", [{}])[0].get("text", "")
                        if text_content:
                            logger.info("Successfully received Gemini chat response.")
                            return text_content.strip()
            except Exception as e:
                logger.error(f"Gemini chat API call failed: {e}. Falling back to mock chat response.", exc_info=True)

        last_message = messages[-1]["content"] if messages else ""
        return self._generate_mock_chat_response(last_message, messages)

    def _generate_mock_chat_response(self, last_message: str, messages: list) -> str:
        q_lower = (last_message or "").lower()
        if "budget" in q_lower or "cost" in q_lower or "price" in q_lower:
            return "I can help customize your budget! Roamio allows you to set your budget per person and choose between Budget, Moderate, and Luxury tiers. Where would you like to travel?"
        elif "hotel" in q_lower or "stay" in q_lower:
            return "We recommend verified stays ranging from boutique homestays to premium luxury resorts tailored to your travel route. Which destination are you planning for?"
        elif "route" in q_lower or "distance" in q_lower or "how to reach" in q_lower:
            return "Roamio automatically computes the fastest road and scenic transit routes between your origin and destination stops. You can view the dynamic map on the Planner page!"
        elif "meghalaya" in q_lower or "shillong" in q_lower:
            return "Meghalaya is a stunning destination! Popular spots include Umiam Lake, Elephant Falls, Cherrapunji's living root bridges, and Dawki's crystal clear river. Would you like a 3-day or 5-day itinerary?"
        elif "himachal" in q_lower or "manali" in q_lower or "shimla" in q_lower:
            return "Himachal Pradesh offers incredible mountain landscapes! Manali and Shimla are top destination hubs with snow viewpoints, paragliding in Solang, and Mall Road heritage. How many days are you planning?"
        else:
            return "Hello! I am Roamio's AI travel consultant. Tell me where you'd like to go, your preferred travel duration, budget, or travel style, and I'll help plan your complete itinerary!"

    async def generate_itinerary(
        self, 
        dest: str, 
        days: int, 
        budget: float, 
        comfort: str, 
        transport: str, 
        place_types: list,
        real_places: list = None
    ) -> dict | None:
        """
        Generates structured JSON itinerary based on user preferences using Gemini 2.5 Flash.
        """
        if not self.api_key or not self.url:
            logger.warning("Gemini Service initialized without an API key. Skipping Gemini generation.")
            return None

        logger.info(f"Gemini Service: Generating itinerary for destination: '{dest}'")
        
        places_context = ""
        if real_places:
            places_context = "\n".join([f"- Name: {p.get('name')}. Summary: {p.get('summary', '')}" for p in real_places[:8]])
        else:
            places_context = "No nearby places context available."

        user_content = f"""
        Generate a {days}-day itinerary for {dest}.
        Number of travelers: 2
        Total Budget: {budget}
        Comfort Level: {comfort}
        Transport Preference: {transport}
        Interests: {", ".join(place_types)}
        
        Here are the REAL-WORLD attractions in/near {dest} that you MUST include and cluster:
        {places_context}
        
        Do not invent other cities or geographic locations outside {dest} and its surroundings.
        """

        payload = {
            "contents": [
                {
                    "parts": [
                        {
                            "text": user_content
                        }
                    ]
                }
            ],
            "systemInstruction": {
                "parts": [
                    {
                        "text": SYSTEM_PROMPT_ITINERARY
                    }
                ]
            },
            "generationConfig": {
                "responseMimeType": "application/json"
            }
        }

        try:
            req_data = json.dumps(payload).encode("utf-8")
            req = urllib.request.Request(
                self.url,
                data=req_data,
                headers={
                    "Content-Type": "application/json"
                },
                method="POST"
            )
            
            ctx = ssl.create_default_context()
            ctx.check_hostname = False
            ctx.verify_mode = ssl.CERT_NONE

            logger.info("Sending request to Gemini API...")
            with urllib.request.urlopen(req, context=ctx, timeout=45) as response:
                resp_data = json.loads(response.read().decode("utf-8"))
                
                candidates = resp_data.get("candidates", [])
                if not candidates:
                    logger.error(f"Gemini API returned no candidates: {resp_data}")
                    return None
                
                text_content = candidates[0].get("content", {}).get("parts", [{}])[0].get("text", "")
                if not text_content:
                    logger.error(f"Gemini API returned empty text: {resp_data}")
                    return None
                
                cleaned_text = text_content.strip()
                if cleaned_text.startswith("```json"):
                    cleaned_text = cleaned_text[7:]
                elif cleaned_text.startswith("```"):
                    cleaned_text = cleaned_text[3:]
                if cleaned_text.endswith("```"):
                    cleaned_text = cleaned_text[:-3]
                cleaned_text = cleaned_text.strip()
                
                itinerary_data = json.loads(cleaned_text)
                logger.info("Successfully received and parsed Gemini API itinerary response.")
                return itinerary_data
                
        except Exception as e:
            logger.error(f"Gemini API call failed: {e}", exc_info=True)
            return None

    async def extract_search_preferences(self, query: str) -> dict:
        """
        Extracts structured travel preferences from natural language query using Gemini 2.5 Flash.
        """
        if not self.api_key or not self.url:
            logger.info(f"GEMINI_API_KEY not set. Using resilient rule-based preference extraction for query: '{query}'")
            return self._fallback_extract_preferences(query)

        logger.info(f"Gemini Service: Extracting travel preferences for query: '{query}'")

        user_content = f'User Travel Query: "{query.strip()}"\nExtract structured preferences into JSON:'

        payload = {
            "contents": [
                {
                    "parts": [
                        {
                            "text": user_content
                        }
                    ]
                }
            ],
            "systemInstruction": {
                "parts": [
                    {
                        "text": SYSTEM_PROMPT_SEARCH_EXTRACTION
                    }
                ]
            },
            "generationConfig": {
                "responseMimeType": "application/json"
            }
        }

        try:
            req_data = json.dumps(payload).encode("utf-8")
            req = urllib.request.Request(
                self.url,
                data=req_data,
                headers={
                    "Content-Type": "application/json"
                },
                method="POST"
            )
            
            ctx = ssl.create_default_context()
            ctx.check_hostname = False
            ctx.verify_mode = ssl.CERT_NONE

            logger.info("Sending natural-language extraction request to Gemini API...")
            with urllib.request.urlopen(req, context=ctx, timeout=30) as response:
                resp_data = json.loads(response.read().decode("utf-8"))
                
                candidates = resp_data.get("candidates", [])
                if not candidates:
                    logger.error(f"Gemini API returned no candidates: {resp_data}")
                    return self._fallback_extract_preferences(query)
                
                text_content = candidates[0].get("content", {}).get("parts", [{}])[0].get("text", "")
                if not text_content:
                    logger.error(f"Gemini API returned empty text: {resp_data}")
                    return self._fallback_extract_preferences(query)
                
                cleaned_text = text_content.strip()
                if cleaned_text.startswith("```json"):
                    cleaned_text = cleaned_text[7:]
                elif cleaned_text.startswith("```"):
                    cleaned_text = cleaned_text[3:]
                if cleaned_text.endswith("```"):
                    cleaned_text = cleaned_text[:-3]
                cleaned_text = cleaned_text.strip()
                
                extracted_data = json.loads(cleaned_text)
                if isinstance(extracted_data, dict):
                    if extracted_data.get("destination"):
                        extracted_data["destination"] = _clean_geographic_entity(extracted_data["destination"])
                    if extracted_data.get("region"):
                        extracted_data["region"] = _clean_geographic_entity(extracted_data["region"])
                    if extracted_data.get("location"):
                        extracted_data["location"] = _clean_geographic_entity(extracted_data["location"])
                logger.info(f"Successfully extracted search preferences from Gemini: {extracted_data}")
                return extracted_data
                
        except Exception as e:
            logger.error(f"Gemini search preference extraction failed: {e}", exc_info=True)
            logger.info(f"Falling back to resilient rule-based preference extraction for query: '{query}'")
            return self._fallback_extract_preferences(query)

    def _fallback_extract_preferences(self, query: str) -> dict:
        q = query.strip()
        lower = q.lower()

        word_to_num = {
            "one": 1, "two": 2, "three": 3, "four": 4, "five": 5,
            "six": 6, "seven": 7, "eight": 8, "nine": 9, "ten": 10
        }

        known_regions = [
            "South India", "North India", "East India", "West India",
            "Northeast India", "Western Ghats", "Eastern Ghats", "Himalayas"
        ]

        destination = None
        region = None
        location = None

        dest_match = re.search(
            r'\b(?:destinations?\s+(?:in|to|at|near|around|close\s+to)|places?\s+(?:in|to|at|near|around|close\s+to)|trip\s+(?:in|to|around|near)|nearby|close\s+to|around|in|to|visit)\s+([a-zA-Z\s]+?)(?:\s+(?:for|from|with|and|\d|$)|$)',
            q,
            re.I
        )
        if dest_match:
            cand = dest_match.group(1).strip()
            cleaned_cand = _clean_geographic_entity(cand)
            if cleaned_cand:
                destination = cleaned_cand

        loc_match = re.search(r'\b(?:from|starting\s+(?:at|from))\s+([a-zA-Z\s]+?)(?:\s+(?:for|in|to|with|and|\d|$)|$)', q, re.I)
        if loc_match:
            cand = loc_match.group(1).strip()
            cleaned_cand = _clean_geographic_entity(cand)
            if cleaned_cand:
                location = cleaned_cand

        if not destination and not location:
            clean_q = re.sub(r'^(?:find|show|give|search|explore|good|best|some|top|places?|destinations?|a\s+place|\s)+', '', lower, flags=re.I).strip()
            cleaned_dest = _clean_geographic_entity(clean_q)
            if cleaned_dest:
                destination = cleaned_dest

        destination = _clean_geographic_entity(destination)
        location = _clean_geographic_entity(location)

        if destination:
            for kr in known_regions:
                if kr.lower() == destination.lower():
                    destination = kr
                    region = kr
                    break

        if location:
            for kr in known_regions:
                if kr.lower() == location.lower():
                    destination = kr
                    region = kr
                    location = None
                    break

        travellers = None
        if re.search(r'\bsolo\b', lower):
            travellers = 1
        elif re.search(r'\bcouple\b', lower):
            travellers = 2
        else:
            t_match = re.search(r'(?:for\s+)?(\d+|one|two|three|four|five|six|seven|eight|nine|ten)\s*(?:travellers?|travelers?|people|persons?|guests?|adults?)', lower)
            if t_match:
                val = t_match.group(1).lower()
                travellers = word_to_num.get(val, int(val) if val.isdigit() else 2)

        duration = None
        if re.search(r'\bweekend\b', lower):
            duration = 2
        else:
            d_match = re.search(r'(?:for\s+)?(\d+|one|two|three|four|five|six|seven|eight|nine|ten)\s*(?:days?|nights?|d\b)', lower)
            if d_match:
                val = d_match.group(1).lower()
                duration = word_to_num.get(val, int(val) if val.isdigit() else 2)

        budget = None
        b_match = re.search(r'(?:under|below|within|budget\s+(?:of)?|₹|rs\.?|inr)\s*(\d+[\d,]*)(?:\s*(?:k|thousand))?', lower)
        if b_match:
            raw_b = b_match.group(1).replace(',', '')
            val = float(raw_b)
            if 'k' in b_match.group(0).lower() or 'thousand' in b_match.group(0).lower():
                val *= 1000
            budget = val

        preferences = []
        if re.search(r'\badventure\b', lower):
            preferences.append("adventure")
        if re.search(r'\b(?:quiet|peaceful|secluded)\b', lower):
            preferences.append("quiet")
        if re.search(r'\b(?:relaxing|relaxation)\b', lower):
            preferences.append("relaxation")
        if re.search(r'\bnature\b', lower):
            preferences.append("nature")
        if re.search(r'\bspiritual\b', lower):
            preferences.append("spiritual")
        if re.search(r'\bromantic\b', lower):
            preferences.append("romantic")

        return {
            "intent": "discover_destinations",
            "destination": destination,
            "region": region,
            "location": location,
            "duration": duration,
            "travellers": travellers,
            "budget": budget,
            "travel_mode": None,
            "preferences": preferences
        }

File: frontend/src/services/ai/geminiService.js
import { envConfig } from '../envConfig';
import { parser } from './parser';
import {
  SYSTEM_PROMPT_CHAT,
  SYSTEM_PROMPT_ITINERARY,
  SYSTEM_PROMPT_ATTRACTIONS,
  SYSTEM_PROMPT_TRAVEL_TIPS,
  SYSTEM_PROMPT_BLUEPRINT,
  getItineraryUserPrompt,
  getAttractionsUserPrompt,
  getTravelTipsUserPrompt,
  getBlueprintUserPrompt
} from './prompts';
import {
  SYSTEM_PROMPT_FILTER_EXTRACTION,
  validateExtractedPreferences,
  extractPreferencesLocalFallback,
  applyPreferencesToFilterState,
  createEmptyPreferences,
} from './filterExtractor';

const GEMINI_API_URL = 'https://generativelanguage.googleapis.com/v1beta/models/gemini-1.5-flash:generateContent';

export const geminiService = {
  /**
   * Natural Language Search -> FastAPI Backend -> Structured Roamio Filter Values
   * Delegates preference extraction to the FastAPI backend (/api/search/extract-preferences)
   * which executes the LLM in Python with full context.
   */
  async extractPreferences(query) {
    if (!query || typeof query !== 'string' || !query.trim()) {
      return createEmptyPreferences();
    }

    console.log(`[geminiService] Sending natural-language query to FastAPI backend: "${query}"`);

    try {
      const response = await fetch('/api/search/extract-preferences', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ query: query.trim() }),
      });

      if (!response || !response.ok) {
        const errDetail = response ? await response.text() : 'No response';
        throw new Error(`FastAPI search extraction failed (${response ? response.status : 'Network error'}): ${errDetail}`);
      }

      const backendData = await response.json();
      console.log("[geminiService] Successfully received structured response from FastAPI backend:", backendData);

      // Map backend response format into Roamio filter structure
      const mapped = {
        location: backendData.location || null,
        destination: backendData.destination || backendData.region || null,
        region: backendData.region || null,
        intent: backendData.intent || 'discover_destinations',
        travellers: backendData.travellers || null,
        duration_days: backendData.duration || null,
        budget_per_person: backendData.budget || null,
        max_travel_time: backendData.max_travel_time || null,
        travel_mode: backendData.travel_mode || null,
        trip_type: backendData.preferences || backendData.trip_type || null,
        accommodation: backendData.accommodation || null,
        crowd_level: backendData.crowd_level || null,
      };

      const validated = validateExtractedPreferences(mapped);
      console.log("[geminiService] Validated preferences from backend:", validated);
      return validated;
    } catch (error) {
      console.error("[geminiService] FastAPI backend search extraction failed:", error);
      console.warn("[geminiService] Falling back to local client extraction to ensure search continuity");
      const fallback = extractPreferencesLocalFallback(query);
      return validateExtractedPreferences(fallback);
    }
  },

  applyPreferencesToFilterState,

  /**
   * Helper to query Gemini API via fetch
   */
  async _callGemini(systemInstruction, prompt) {
    const key = envConfig.GEMINI_API_KEY;
    if (!key) {
      throw new Error("Missing Gemini API Key");
    }

    const url = `${GEMINI_API_URL}?key=${key}`;
    const requestBody = {
      contents: [
        {
          parts: [{ text: prompt }]
        }
      ],
      systemInstruction: {
        parts: [{ text: systemInstruction }]
      },
      generationConfig: {
        responseMimeType: 'application/json',
        temperature: 0.4
      }
    };

    const response = await fetch(url, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(requestBody)
    });

    if (!response.ok) {
      const errorText = await response.text();
      throw new Error(`Gemini API request failed: ${response.status} - ${errorText}`);
    }

    const data = await response.json();
    const textContent = data.candidates?.[0]?.content?.parts?.[0]?.text;
    
    if (!textContent) {
      throw new Error("No text content returned from Gemini API");
    }

    return textContent;
  },

  /**
   * Generates a complete trip itinerary plan.
   */
  async generateTripPlan(destination, duration = 4, budget = 20000, preferences = {}) {
    console.log(`[geminiService] Generating trip plan for ${destination} (${duration} days, Budget: ${budget})`);
    
    if (!envConfig.GEMINI_API_KEY) {
      console.warn("[geminiService] No Gemini API key configuration found. Serving high-fidelity mock trip plan.");
      return this._getMockTripPlan(destination, duration, budget, preferences);
    }

    try {
      const prompt = getItineraryUserPrompt(destination, duration, budget, preferences);
      const rawResponse = await this._callGemini(SYSTEM_PROMPT_ITINERARY, prompt);
      const parsed = parser.parseJSON(rawResponse);
      if (!parsed) {
        throw new Error("Parsed JSON is null");
      }
      return parsed;
    } catch (error) {
      console.error("[geminiService] generateTripPlan failed, falling back to mock:", error);
      return this._getMockTripPlan(destination, duration, budget, preferences);
    }
  },

  /**
   * Generates 5 nearby attractions.
   */
  async generateAttractions(destination) {
    console.log(`[geminiService] Generating attractions for ${destination}`);

    if (!envConfig.GEMINI_API_KEY) {
      console.warn("[geminiService] No Gemini API key. Serving mock attractions.");
      return this._getMockAttractions(destination);
    }

    try {
      const prompt = getAttractionsUserPrompt(destination);
      const rawResponse = await this._callGemini(SYSTEM_PROMPT_ATTRACTIONS, prompt);
      const parsed = parser.parseJSON(rawResponse);
      if (!parsed) {
        throw new Error("Parsed attractions array is null");
      }
      return parsed;
    } catch (error) {
      console.error("[geminiService] generateAttractions failed, falling back to mock:", error);
      return this._getMockAttractions(destination);
    }
  },

  /**
   * Generates localized travel tips.
   */
  async generateTravelTips(destination) {
    console.log(`[geminiService] Generating travel tips for ${destination}`);

    if (!envConfig.GEMINI_API_KEY) {
      console.warn("[geminiService] No Gemini API key. Serving mock travel tips.");
      return this._getMockTravelTips(destination);
    }

    try {
      const prompt = getTravelTipsUserPrompt(destination);
      const rawResponse = await this._callGemini(SYSTEM_PROMPT_TRAVEL_TIPS, prompt);
      const parsed = parser.parseJSON(rawResponse);
      if (!parsed) {
        throw new Error("Parsed travel tips is null");
      }
      return parsed;
    } catch (error) {
      console.error("[geminiService] generateTravelTips failed, falling back to mock:", error);
      return this._getMockTravelTips(destination);
    }
  },

  /**
   * Generates detailed blueprint text.
   */
  async generateBlueprintContent(destination, tripPlan) {
    console.log(`[geminiService] Generating blueprint for ${destination}`);

    if (!envConfig.GEMINI_API_KEY) {
      console.warn("[geminiService] No Gemini API key. Serving mock blueprint.");
      return this._getMockBlueprint(destination, tripPlan);
    }

    try {
      const prompt = getBlueprintUserPrompt(destination, tripPlan);
      const rawResponse = await this._callGemini(SYSTEM_PROMPT_BLUEPRINT, prompt);
      const parsed = parser.parseJSON(rawResponse);
      if (!parsed) {
        throw new Error("Parsed blueprint is null");
      }
      return parsed;
    } catch (error) {
      console.error("[geminiService] generateBlueprintContent failed, falling back to mock:", error);
      return this._getMockBlueprint(destination, tripPlan);
    }
  },

  /**
   * Chat bot assistant response generation.
   */
  async generateChatResponse(messageHistory, context = {}) {
    console.log("[geminiService] Generating chat response");

    if (!envConfig.GEMINI_API_KEY) {
      console.warn("[geminiService] No Gemini API key. Serving mock chat response.");
      return this._getMockChatResponse(messageHistory, context);
    }

    try {
      const key = envConfig.GEMINI_API_KEY;
      const url = `${GEMINI_API_URL}?key=${key}`;
      
      // Structure message history for Gemini API
      // Translate 'user'/'assistant' roles to 'user'/'model' expected by Gemini
      const contents = messageHistory.map(m => ({
        role: m.role === 'assistant' ? 'model' : 'user',
        parts: [{ text: m.content }]
      }));

      // Append trip context to system instruction
      const contextualSystemInstruction = `${SYSTEM_PROMPT_CHAT}\n\nCURRENT TRIP CONTEXT:\n${JSON.stringify(context, null, 2)}`;

      const requestBody = {
        contents,
        systemInstruction: {
          parts: [{ text: contextualSystemInstruction }]
        },
        generationConfig: {
          temperature: 0.7,
          maxOutputTokens: 1000
        }
      };

      const response = await fetch(url, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify(requestBody)
      });

      if (!response.ok) {
        throw new Error(`Gemini Chat API returned status ${response.status}`);
      }

      const data = await response.json();
      const reply = data.candidates?.[0]?.content?.parts?.[0]?.text || "I'm having trouble thinking of a response. Please try again.";
      return {
        role: "assistant",
        content: reply
      };
    } catch (error) {
      console.error("[geminiService] generateChatResponse failed, falling back to mock:", error);
      return this._getMockChatResponse(messageHistory, context);
    }
  },

  // ==========================================
  // HIGH FIDELITY MOCK GENERATORS
  // ==========================================

  _getMockTripPlan(dest, duration, budget, preferences) {
    const travelers = preferences.travelers || 2;
    const comfort = preferences.comfortLevel || 'moderate';
    
    // Dynamic costs calculations
    const transport_cost = Math.round(budget * 0.25);
    const accommodation_cost = Math.round(budget * 0.35);
    const activities_cost = Math.round(budget * 0.20);
    const food_cost = Math.round(budget * 0.15);
    const remaining_savings = Math.round(budget * 0.05);

    return {
      travelers: parseInt(travelers),
      budget: parseFloat(budget),
      destination: dest,
      currency: "INR",
      total_days: parseInt(duration),
      comfort_level: comfort,
      transport_preference: preferences.transportPreference || 'fastest',
      budget_utilization: {
        transport_cost,
        accommodation_cost,
        activities_cost,
        food_cost,
        remaining_savings
      },
      regions: [
        {
          region_name: `${dest} Valley`,
          region_image_query: `${dest} nature landscape`,
          quick_summary: `Explore the vibrant heart and scenic natural corridors of ${dest}.`,
          transport: "private cab",
          hotel_type: comfort === 'budget' ? 'Cozy Homestay' : comfort === 'moderate' ? 'Premium Cottage' : 'Luxury Resort',
          estimated_cost: Math.round(budget * 0.6),
          places: [`${dest} Overlook`, `${dest} Historic Market`, `${dest} Nature Reserve`],
          food_recommendations: [`Traditional ${dest} Platter`, "Highland Brew Cafe"],
          activities: ["Scenic Photography", "Local Market Walks"],
          timeline: Array.from({ length: duration }).map((_, i) => ({
            day: i + 1,
            title: `Day ${i + 1}: Discovering ${dest} Highlights`,
            activities: [
              {
                time: "09:00 AM",
                activity: `Scenic drive to ${dest} central landmarks and valleys.`,
                cost: 800
              },
              {
                time: "01:00 PM",
                activity: "Lunch and tasting at local traditional food hub.",
                cost: 400
              },
              {
                time: "03:30 PM",
                activity: `Hike and nature walk across ${dest} trails.`,
                cost: 200
              }
            ]
          }))
        }
      ],
      nearby_attractions: this._getMockAttractions(dest)
    };
  },

  _getMockAttractions(dest) {
    return [
      {
        name: `${dest} Scenic Falls`,
        summary: "Breathtaking multi-tiered waterfall nestled in green forests.",
        visit_duration: "2 Hours",
        local_cost: 150.0,
        description: `The highest and most famous waterfall in ${dest}, offering stunning viewpoints, trekking trails, and local handicraft stalls.`,
        best_time: "October to April",
        tips: "Visit in the early morning to capture photos with the fog clearing.",
        distance: "8 km from center",
        local_transport: "Local Taxi",
        highlights: ["Three-tier view", "Trekking stairs", "Local tea stalls"],
        quick_facts: {
          "Type": "Waterfall",
          "Rating": "4.8 / 5.0"
        },
        nearby_activities: ["Photography", "Fringe Hiking"]
      },
      {
        name: `${dest} Heritage Village`,
        summary: "Experience authentic traditions and clean green living.",
        visit_duration: "3 Hours",
        local_cost: 50.0,
        description: "A beautifully preserved heritage village showcasing clean eco-friendly lifestyles, traditional architecture, and delicious home-cooked tribal meals.",
        best_time: "All year round",
        tips: "Hire a local guide to learn the deep folklore and stories.",
        distance: "15 km from center",
        local_transport: "Shared Cab",
        highlights: ["Eco-friendly homes", "Living architecture", "Traditional crafts"],
        quick_facts: {
          "Type": "Cultural Site",
          "Rating": "4.7 / 5.0"
        },
        nearby_activities: ["Village Tour", "Traditional Lunch"]
      },
      {
        name: `${dest} Clouds Valley`,
        summary: "A spectacular viewpoint suspended over sea of clouds.",
        visit_duration: "1.5 Hours",
        local_cost: 0.0,
        description: "Witness the magical early morning mist rising from deep valley corridors, surrounding you in white clouds.",
        best_time: "Monsoon and Winter",
        tips: "Reach by 5:00 AM for the spectacular sunrise display.",
        distance: "22 km from center",
        local_transport: "Private Car",
        highlights: ["Cloud ocean views", "Sunrise panorama", "Mountain range"],
        quick_facts: {
          "Type": "Viewpoint",
          "Rating": "4.9 / 5.0"
        },
        nearby_activities: ["Sunrise Watch", "Landscape Photography"]
      },
      {
        name: `${dest} Crystal River`,
        summary: "Boating on perfectly clear water.",
        visit_duration: "2.5 Hours",
        local_cost: 500.0,
        description: "Famous for its mirror-like transparent water where boats look like they are floating in mid-air. Perfect for boating, camping, and kayaking.",
        best_time: "November to March",
        tips: "Boating is best on sunny days for maximum transparency.",
        distance: "30 km from center",
        local_transport: "Prepaid Cab",
        highlights: ["Crystal clear water", "Boating tour", "Suspension bridge"],
        quick_facts: {
          "Type": "River / Adventure",
          "Rating": "4.8 / 5.0"
        },
        nearby_activities: ["Kayaking", "River Camping", "Boating"]
      },
      {
        name: `${dest} Sacred Forest`,
        summary: "An ancient forest preserved by traditional laws.",
        visit_duration: "2 Hours",
        local_cost: 100.0,
        description: "A mysterious and dense forest where not even a leaf or twig is allowed to be taken out. Rich in ancient monoliths and exotic orchids.",
        best_time: "October to May",
        tips: "Respect local beliefs: do not carry anything out from the forest.",
        distance: "12 km from center",
        local_transport: "Taxi",
        highlights: ["Ancient monoliths", "Rare orchids", "Centuries old trees"],
        quick_facts: {
          "Type": "Nature / Sacred Grove",
          "Rating": "4.6 / 5.0"
        },
        nearby_activities: ["Guided walk", "Flora spotting"]
      }
    ];
  },

  _getMockTravelTips(dest) {
    return {
      logistics: {
        best_time: "October to April is ideal with pleasant temperatures.",
        packing_essentials: ["Comfortable hiking shoes", "Windbreaker jacket", "Umbrella (weather is unpredictable)", "Refillable water bottle"],
        connectivity: "Jio and Airtel offer good 4G network coverage in main hubs. Valley signals can be intermittent.",
        local_commute: "Private taxi packages are the most comfortable and reliable options."
      },
      cultural_norms: {
        dress_code: "Modest casual clothing. Cover knees and shoulders when visiting community shrines or holy sites.",
        social_etiquettes: ["Always ask permission before photographing locals", "Keep the environment clean (littering is strictly frowned upon)"],
        photography_rules: "Permitted everywhere except inside sacred inner shrines."
      },
      safety: {
        emergency_contacts: {
          police: "112 / 100",
          tourist_helpline: "+91-888-TRIP-HELP"
        },
        health_guidelines: "Carry motion sickness pills for winding mountain roads. Drink bottled or filtered water.",
        scam_alerts: "Agree on taxi fares before boarding. Avoid unlicensed guides near main viewpoints."
      }
    };
  },

  _getMockBlueprint(dest, tripPlan) {
    return {
      title: `The Ultimate ${dest} Explorer Guide`,
      subtitle: `A curated Blueprint for a ${tripPlan?.total_days || 4}-Day trip`,
      packing_list: [
        "Robust walking shoes for slopes",
        "Rain protection / Umbrella",
        "Motion sickness meds",
        "Layered clothing / Light jacket"
      ],
      route_highlights: [
        `Traverse beautiful regions of ${dest} valley`,
        "Explore majestic local waterfalls and viewpoints",
        "Discover historic tribal villages and pristine environments"
      ],
      must_try_dishes: [
        `Traditional local ${dest} stew`,
        "Smoked bamboo shoot curry",
        "Signature organic ginger tea"
      ],
      local_words: [
        { word: "Thank you", translation: "Khublei / Dhanyavaad", pronunciation: "Koob-lay / Dan-ya-vaad" },
        { word: "How much?", translation: "Katno? / Kitna?", pronunciation: "Kat-no / Kit-na" },
        { word: "Hello", translation: "Kumno / Namaste", pronunciation: "Koom-no / Na-mas-tay" }
      ],
      emergency_contacts: [
        "Local Tourist Office: +91-XXX-XXXX",
        "National Emergency Helpline: 112"
      ],
      checklist: [
        "Double check weather forecasts",
        "Pre-book local taxi transport option",
        "Download offline maps of the valley",
        "Confirm homestay check-in timings"
      ]
    };
  },

  _getMockChatResponse(history, context) {
    const lastMsg = history[history.length - 1]?.content?.toLowerCase() || '';
    let response = "I'm here to help you coordinate your Antigravity travel plans! Let me know if you need help with destinations, budget adjustments, or generating your itinerary.";

    if (lastMsg.includes("budget")) {
      response = `For a trip to ${context.destination || "your destination"}, we can optimize your expenses. Would you like to reduce the current budget or find cheaper homestays?`;
    } else if (lastMsg.includes("waterfall") || lastMsg.includes("attraction")) {
      response = `I highly recommend checking out some scenic local waterfalls and historic villages in ${context.destination || "your destination"}. You can add them to your travel cart!`;
    } else if (lastMsg.includes("how") || lastMsg.includes("help") || lastMsg.includes("guide")) {
      response = "To create a trip, enter your destination and budget in the form on the left, select your interests, then click 'Generate Plan'. You can customize stops or download your Blueprint Guide anytime!";
    }

    return {
      role: "assistant",
      content: response
    };
  }
};
